### DISCUSSION TITLE
Robust MPC errors for solving a MIMO linear system: Cannot get rid of nonlinear uncertainty in uncertain constraint

### DISCUSSION THREAD

**Discussion by guixianqu** [yalmip/YALMIP#952] [Problems using YALMIP] [answered]
Hi Johan and everyone,

I am a beginner in optimization and semidefinite programming and got trouble running a robust MPC demo for solving a multi-input-multi-output (MIMO) linear system. To be specific, the MIMO linear system in the presence of some additive disturbances is as follows X(k+1) = AX(k) + Bu(k) + Ew(k), where A is of 15*15, B and E are of 15*5. That is, the state dimension is 15, while both the control and the disturbance dimensions are equal to 5. In addition, A is a full real matrix.

I have followed the robust MPC example provided by the official website, i.e., "[Model predictive control - robust solutions](https://yalmip.github.io/example/robustmpc/)
[MIMO_Robust_MPC_Demo.zip](https://github.com/yalmip/YALMIP/files/6873355/MIMO_Robust_MPC_Demo.zip)
", to solve the above problem. But, unfortunately, I got the following error when running the program, saying that "*Cannot get rid of nonlinear uncertainty in uncertain constraint*":

```matlab
Error using filter_eliminatation (line 15)
Cannot get rid of nonlinear uncertainty in uncertain constraint

Error in robustify (line 147)
[UncertainModel.F_xw,F_elimination] =
filter_eliminatation(UncertainModel.F_xw,w,1,ops);

Error in optimizer (line 184)
    [Constraints,Objective,failure] =
    robustify(Constraints,Objective,options);
```

 The following is my MATLAB code, which is quite similar to the official example:

```matlab
%% 
clear
close all

yalmip('clear');

pack;
clc;


%% 
data = load('MIMO_Example.mat');

% The system parameters, i.e., A, B, E, Q, R
A = data.A; B = data.B; E = data.E; Q = data.Q; R = data.R;

% The bounds on the state and the control
X_lb = data.state_lb; X_ub = data.state_ub;
u_lb = data.control_lb; u_ub = data.control_ub;

% The prediction horizon
N = 3;

% The state dimension
Nx = 15;

% The control dimension
Nu = 5;

% The disturbance dimension
Nw = 5;

%%
% Define the initial state variable
X_0 = sdpvar(Nx, 1);

% Define the global control and uncertainty variables
U = sdpvar(repmat(Nu,1,N), repmat(1, 1, N));
W = sdpvar(repmat(Nu,1, N), repmat(1, 1, N));

% Define the objective and constraints
constraints = [];
objective = 0;

X_k = X_0;
for k = 1:N    
    % Do prediction
    X_k = A*X_k + B*U{k} + E*W{k};
    
    % The objective
    objective = objective + ...
        0.5.*((X_k')*Q*X_k + (U{k}')*R*U{k});
    
    % Set constraints on the control at current instant k
    constraints = [constraints, ...
        u_lb <= U{k} <= u_ub, ...
        -1 <= W{k} <= 1, uncertain(W{k})];

    % Set constraints on the predicted state at current instant k
    constraints = [constraints, X_lb <= X_k <= X_ub];
end

ops = sdpsettings;
% ops = sdpsettings('solver', 'mosek', ...
%     'verbose', 1, ...
%     'debug', 1, ...
%     'showprogress', 1);

controller = optimizer(constraints, objective, ops, X_0, U{1});


%% 
simulation_horizon = 25;

state_trajectory = zeros(Nx, simulation_horizon + 1);
state_trajectory(:, 1) = data.initial_state;

control_trajectory = zeros(Nu, simulation_horizon);

for k = 2 : simulation_horizon + 1
    % Calculate the control signal
    control_trajectory(:, k-1) = controller{state_trajectory(:, k-1)};
    
    % Implement the control
    state_trajectory(:, k) = A*state_trajectory(:, k-1) + B*control_trajectory(:, k-1) + E*(-1 + 2.*rand(Nu, 1));
    
end


%%
figure;
plot(state_trajectory([1, 4, 7, 10, 13],:)');

figure;
plot(state_trajectory([2, 5, 8, 11, 14],:)');

figure;
plot(state_trajectory([3, 6, 9, 12, 15],:)');

figure;
plot(control_trajectory');

```

I have also attached the MATLAB source file and data for review and error reproducibility. 

Besides, I would also like to solve my MPC model by adopting the approximate closed-loop minimax solution (i.e., to parameterize future inputs as affine functions of past disturbances) as illustrated on the official website. But, I have no idea to write the corresponding MATLAB program simply by following the official demo, since the example given in the official demo aims at a single-input-single-output (SISO) system whose control and disturbance dimensions are only one (while my targeted system with multiple control inputs and disturbances is more complicated).

Could you do me a favor? Thanks a lot!

Best regards,

Guixian

**Comment by johanlofberg** [comment, marked as answer]
The robust layer does not have any strategy to deal with uncertain quadratics, there is a reason the example uses an L-inf performance measure

To extend to multivariable, you will have to change dimensions so they make sense. Once you have the dimensions of x,u and w it should be obvious how to update the code (the only tricky part is defining an L of suitable causal sparsity pattern, should be something like sdpvar(N*nu,N*nw).*kron(tril(ones(N),-1),ones(nu,nw)))

**Reply by guixianqu** [reply]
Thank you, Johan. I will revise my code according to your suggestion. Thanks!

**Reply by WAN-OPA** [reply]
Hello, Elder Brother. How did you solve this problem
